### ISSUE TITLE
Separate machines across different networks?

### ISSUE THREAD

**Issue by galalmounir** [herlesupreeth/docker_open5gs#231]
What is the best way to run the core dockers on one machine A on network A then run the ENB on a separate machine B on network B and have them communicate?

What I tried was putting Machine A's IP in `DOCKER_HOST_IP`, `SGWU_ADVERTISE_IP` and `UPF_ADVERTISE_IP`. 
I also uncommented the lines in the docker files. 
Finally, I port forwarded the two ports that are open in the docker files. 

The first issue I faced was being unable to run the srsenb.yaml docker-compose on Machine B, because the virtual docker network (external) is not available on that machine. I tried to manually create the virtual network. But was unable to get machine B (ENB) to connect to Machine A's core.

Any ideas?

**Comment by herlesupreeth** [comment]
Hey, if you are running eNB on an another machine then you would need to change the networking to host mode (rather than external) of srsenb.yaml docker-compose file so that there is no NATing done by the docker network.

i.e. remove the networks sections and add under srsenb section in that docker-compose file the following

`network_mode: host`

**Comment by galalmounir** [comment]
Thanks! 

I tried this and I got the following error:
```
srsenb    | bind(): Cannot assign requested address
srsenb    | Failed to bind on address 172.22.0.22, port 2152: Cannot assign requested address
srsenb    | Error initializing EUTRA stack.
```

This is what my srsenb.yaml looks like:
```
version: '3'
services:
  srsenb:
    image: galalhassan/docker_srslte
    container_name: srsenb
    stdin_open: true
    tty: true
    privileged: true
    devices:
      - "/dev/bus"
    volumes:
      - /dev/serial:/dev/serial:ro
      - /dev/bus/usb:/dev/bus/usb:ro
      - ./srslte:/mnt/srslte
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    environment:
      - COMPONENT_NAME=enb
    expose:
      - "36412/sctp"
      - "2152/udp"
    network_mode: "host"
```

**Comment by herlesupreeth** [comment]
Ah my bad, you need to set the SRS_ENB_IP to the host IP of the machine running it (rather than 172.22.022)

**Comment by herlesupreeth** [comment]
> I am assuming this would be the public IP not the local IP of the machine.

The IP at which eNB is able to reach EPC

> What port would we need to expose on the ENB machine/network?

Technically, its 36412 port for SCTP and 2152 for UDP. But since we are running in network_mode: host , those ports should be exposed across all interface of that machine running eNB

**Comment by galalmounir** [comment]
Got it, let me give it a try and report back.

What's the best way to test the connection between the eNB and EPC? (other than running all components and checking logs)

**Comment by galalmounir** [comment]
I am still unable to get both machines working. 
I am now trying the same setup but on the same local network.

- Machine A
  - running the core containers (EPC)
  - Exposes ports 36412 and 2152
- Machine B
  - running the eNB
  - uses network_mode: host (without network_mode set as host, I am unable to start the enb container)

I am checking if they are able to communicate by checking the logs on the `mme`. I don't see the enb connecting to the mme

**Comment by herlesupreeth** [comment]
Can you please send me a trace take on both the machines (without any filters) when you try to connect eNB to EPC ??

**Comment by s5uishida** [comment]
Hi @galalmounir 

I don't think `UPF_ADVERTISE_IP` should be changed.

**Comment by herlesupreeth** [comment]
@s5uishida ah true, in case of 4G `UPF_ADVERTISE_IP` should not be modified. But it shouldn't affect eNB from connecting to EPC

**Comment by galalmounir** [comment]
@herlesupreeth when you say trace, what's the best tool to use? (Wireshark?)

**Comment by galalmounir** [comment]
Without any filters using Wireshark is too much. Both machines are connected to the internet, plus SSH session and VNC sessions.
I'll filter out the traffic with source/destination IPs. That way we filter out anything going out to the internet or over to my main/dev machine.

**Comment by galalmounir** [comment]
The IPs for the machines are as follows:
- Machine A (EPC) -> 192.168.7.168
- Machine B (eNB) -> 192.168.7.222
- Dev Machine -> 192.168.6.3

I used the following filter on Machine A:
```
(ip.src==192.168.7.168 || ip.addr==192.168.7.222) && not ssh && ip.addr!=192.168.6.3
```

And I used the following filter on Machine B:
```
(ip.src==192.168.7.222 || ip.addr==192.168.7.168) && not ssh && ip.addr!=192.168.6.3
```

All the logs are either MDNS or TCP to my dev machine (192.168.6.3).

This is from Machine A:
<img width="1241" alt="image" src="https://github.com/herlesupreeth/docker_open5gs/assets/2583382/4abb6cbe-d09b-4723-aead-f192f2a02a91">

This is from Machine B:
<img width="946" alt="image" src="https://github.com/herlesupreeth/docker_open5gs/assets/2583382/96ad22c1-c6a5-48de-8a8d-b3350c97cab9">

**Comment by galalmounir** [comment]
I did some digging, and it seems like the sctp port is not exposed on Machine 1. Even though it is correctly mapped by docker. (tried using netcat/socat)
I believe this has to do with me running the EPC stack on a MacOS.

I'll try running the EPC stack on a linux machine in the morning and try again.

**Comment by herlesupreeth** [comment]
> I believe this has to do with me running the EPC stack on a MacOS.

That is it. MAC OS as far as I remember doesnt have libsctp

**Comment by herlesupreeth** [comment]
@s5uishida But that would require compiling open5gs with usrsctp support which is not done in Dockerfile of this repo. Let me know if I am missing anything.

**Comment by galalmounir** [comment]
@s5uishida yes I tried installing `libusrsctp` with no avail. :( 
Might be due to what @herlesupreeth mentioned about building open5gs with usrsctp support. 

@herlesupreeth, do you have a link to rebuilding with usrsctp support?

**Comment by herlesupreeth** [comment]
> @herlesupreeth, do you have a link to rebuilding with usrsctp support?

hmmm, I wouldn't recommend it because usrsctp will be used only by MME for connection to eNB but whereas for diameter side of things freediameter will be used, which still uses libsctp (no other SCTP alternative for freediameter) so you wont have a working setup.

I would suggest running a VM in MAC OS and install docker_open5gs in that.

**Comment by s5uishida** [comment]
Hi @herlesupreeth 

For the time being, I thought it was necessary for the sctp kernel module to be available on Docker Host (MAC).

Sorry if I misunderstood.

**Comment by galalmounir** [comment]
I’ll give that a try shortly and report back. Thanks!

**Comment by galalmounir** [comment]
Okay, so after a few trials and errors. I was able to get it working. 

TLDR; the `MME_IP` in the `.env` file also needs to change to the `DOCKER_HOST_IP`


Details:
I setup an AWS ubuntu VM and opened the firewall to the internet. This way I can rule out any ports that are not open. 
Then I started by changing things one by one in the `.env` and `srsenb.yaml` until I was able to start seeing some traffic on wireshark.
This is the filter used: 
`ip.addr=={AWS_IP} || sctp.port==36412 || udp.port==2152`

I saw a few `INIT` requests going to IP `172.22.0.9`, so after looking into the `.env` file. I found that the `MME_IP` is set to `172.22.0.9`, after changing it to the `{AWS_IP}` (on the machine running the eNB), I started seeing more traffic on wireshark with successful S1 connections and heartbeats.

I'll push a PR to add/update the docs to reflect those requirements.

Thanks again for all the help!

**Comment by galalmounir** [comment]
Resolved with suggestions and readme updated in PR #236 to reflect the suggestions
